File: back/Projeto.py
class Projeto:
    def __init__(self, db):
        self.db = db

    def add_projeto(self, id_avaliacao, nome_projeto, projeto_habilitado, numero_projeto):
        cursor = self.db.conn.cursor(dictionary=True)
        query = """
            INSERT INTO projeto 
            (ID_Avaliacao, Nome_Projeto, Projeto_Habilitado, Numero_Projeto) 
            VALUES (%s, %s, %s, %s)
        """
        try:
            cursor.execute(query, (id_avaliacao, nome_projeto, projeto_habilitado, numero_projeto))
            self.db.conn.commit()  # Confirma a transação
            value = cursor.lastrowid  # Obtém o ID do último projeto inserido
            return value
        except Exception as e:
            self.db.conn.rollback()  # Desfaz a operação em caso de erro
            print(f"Erro ao adicionar projeto: {e}")
            raise e
        finally:
            cursor.close()
    def get_projetos_by_id_avaliacao(self, id_avaliacao):
        cursor = self.db.conn.cursor(dictionary=True)
        query_projetos = "SELECT * FROM projeto WHERE ID_Avaliacao = %s"
        
        cursor.execute(query_projetos, (id_avaliacao,))
        projetos = cursor.fetchall()

        projetos_com_documentos = []
        for projeto in projetos:
            query_documentos = """
            SELECT d.ID, d.Caminho_Arquivo, d.Nome_Arquivo, d.ID_Projeto 
            FROM documento d 
            WHERE d.ID_Projeto = %s
            """
            cursor.execute(query_documentos, (projeto['ID'],))
            documentos = cursor.fetchall()
            projetos_com_documentos.append({
                'ID': projeto['ID'],
                'ID_Avaliacao': projeto['ID_Avaliacao'],
                'Projeto_Habilitado': projeto['Projeto_Habilitado'],
                'Numero_Projeto': projeto['Numero_Projeto'],
                'Nome_Projeto': projeto['Nome_Projeto'],
                'Documentos': [
                    {
                        'ID': doc['ID'],
                        'Caminho_Arquivo': doc['Caminho_Arquivo'],
                        'Nome_Arquivo': doc['Nome_Arquivo'],
                        'ID_Projeto': doc['ID_Projeto']
                    } for doc in documentos
                ]
            })
        cursor.close()
        return projetos_com_documentos

    def update_projeto(self, projeto_id, nome_projeto, projeto_habilitado):
        cursor = self.db.conn.cursor(dictionary=True)
        query = """
            UPDATE projeto 
            SET Nome_Projeto = %s, Projeto_Habilitado = %s 
            WHERE ID = %s
        """
        try:
            cursor.execute(query, (nome_projeto, projeto_habilitado, projeto_id))
            self.db.conn.commit()  # Confirma a transação
        except Exception as e:
            self.db.conn.rollback()  # Desfaz a operação em caso de erro
            print(f"Erro ao atualizar projeto: {e}")
            raise e
        finally:
            cursor.close()  # Garante que o cursor será fechado

    def get_next_numero_projeto(self, id_avaliacao):
        cursor = self.db.conn.cursor(dictionary=True)
        query = "SELECT MAX(Numero_Projeto) FROM projeto WHERE ID_Avaliacao = %s"
        cursor.execute(query, (id_avaliacao,))
        result = cursor.fetchone()
        cursor.close()
        print(f"Próximo número de projeto: {result[0]}")
        return (result[0] or 0) + 1

File: backjs/classes/Projeto.js
class Projeto {
    constructor(db) {
        this.db = db;
    }

    async addProjeto(idAvaliacao, nomeProjeto, projetoHabilitado, numeroProjeto) {
        const query = "INSERT INTO projeto (ID_Avaliacao, Nome_Projeto, Projeto_Habilitado, Numero_Projeto) VALUES (?, ?, ?, ?)";
        try {
            await this.db.executeQuery(query, [idAvaliacao, nomeProjeto, projetoHabilitado, numeroProjeto]);
            const lastId = await this.db.getLastInsertId(); // Função que retorna o último ID inserido
            return lastId;
        } catch (err) {
            console.error("Erro ao adicionar projeto:", err);
            throw err;
        }
    }

    async getProjetosByIdAvaliacao(idAvaliacao) {
        const queryProjetos = "SELECT * FROM projeto WHERE ID_Avaliacao = ?";
        try {
            const projetos = await this.db.fetchAll(queryProjetos, [idAvaliacao]);

            const projetosComDocumentos = await Promise.all(projetos.map(async (projeto) => {
                const queryDocumentos = `
                    SELECT d.ID, d.Caminho_Arquivo, d.Nome_Arquivo, d.ID_Projeto 
                    FROM documento d 
                    WHERE d.ID_Projeto = ?
                `;
                const documentos = await this.db.fetchAll(queryDocumentos, [projeto.ID]);
                return {
                    ID: projeto.ID,
                    ID_Avaliacao: projeto.ID_Avaliacao,
                    Projeto_Habilitado: projeto.Projeto_Habilitado,
                    Numero_Projeto: projeto.Numero_Projeto,
                    Nome_Projeto: projeto.Nome_Projeto,
                    Documentos: documentos.map((doc) => ({
                        ID: doc.ID,
                        Caminho_Arquivo: doc.Caminho_Arquivo,
                        Nome_Arquivo: doc.Nome_Arquivo,
                        ID_Projeto: doc.ID_Projeto
                    }))
                };
            }));

            return projetosComDocumentos;
        } catch (err) {
            console.error("Erro ao buscar projetos:", err);
            throw err;
        }
    }

    async updateProjeto(projetoId, nomeProjeto, projetoHabilitado) {
        const query = "UPDATE projeto SET Nome_Projeto = ?, Projeto_Habilitado = ? WHERE ID = ?";
        try {
            await this.db.executeQuery(query, [nomeProjeto, projetoHabilitado, projetoId]);
        } catch (err) {
            console.error("Erro ao atualizar projeto:", err);
            throw err;
        }
    }

    async getNextNumeroProjeto(idAvaliacao) {
        const query = "SELECT MAX(Numero_Projeto) FROM projeto WHERE ID_Avaliacao = ?";
        try {
            const result = await this.db.fetchOne(query, [idAvaliacao]);
            return (result ? result["MAX(Numero_Projeto)"] : 0) + 1;
        } catch (err) {
            console.error("Erro ao obter o próximo número de projeto:", err);
            throw err;
        }
    }
}

module.exports = Projeto;